### PR TITLE
Use distinct headless service for proxies

### LINKED ISSUE #829: Exposing a statefulset creates a proxy statefulset which reuses the application headless service
**Problem statement:**

When a stateful set is exposed as a headless service using skupper, A new stateful set called appname-proxy gets created, but this stateful set reuses the service that is used by the application, This will basically create a problem when you try to access the pod using its identity. See example below.

Redis is deployed as a stateful set, which has a service called redis-headless  created, to be able to access each pod by its name.
$oc get sts
NAME                   READY   AGE
redis-master           1/1     71m
redis-replicas         3/3     71m

$oc get sts redis-master -o yaml
....
spec:
 ....
  **serviceName: redis-headless**
 ....

when i expose redis-master stateful set using skupper as a headless service using the below command a new stateful set called 
redis-headless-proxy  is created and it reuses **redis-headless**  service. Reusing this service name will make it not possible to access pod identity for the stateful set redis-headless-proxy. This will not really matter when we use plain text communication but if we enable end-to-end TLS between applications and if the application expects the certificate to match the hostname of the pod from which the request comes, it creates a problem.

**skupper expose statefulset redis-master  --headless** 

$oc get sts
NAME                   READY   AGE
redis-headless-proxy   1/1     63m
redis-master           1/1     71m
redis-replicas         3/3     71m

$oc get svc
NAME                                    TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)               AGE
redis-headless                          ClusterIP   None             <none>        6379/TCP              72m
redis-master                            ClusterIP   172.21.136.41    <none>        6379/TCP              72m
redis-replicas                          ClusterIP   172.21.241.50    <none>        6379/TCP              72m
skupper                                 ClusterIP   172.21.111.78    <none>        8080/TCP,8081/TCP     78m
skupper-router                          ClusterIP   172.21.131.250   <none>        55671/TCP,45671/TCP   78m
skupper-router-local                    ClusterIP   172.21.197.98    <none>        5671/TCP              78m

$oc get sts redis-headless-proxy -o yaml
spec:
 ..... 
  serviceName: redis-headless
......

ping responses from inside the cluster.
sh-5.0# ping redis-master-0.redis-headless.skupper-local.svc.cluster.local
PING redis-master-0.redis-headless.skupper-local.svc.cluster.local (10.1.0.11) 56(84) bytes of data.
64 bytes from redis-master-0.redis-headless.skupper-local.svc.cluster.local (10.1.0.11): icmp_seq=1 ttl=64 time=0.088 ms
64 bytes from redis-master-0.redis-headless.skupper-local.svc.cluster.local (10.1.0.11): icmp_seq=2 ttl=64 time=0.099 ms

--- redis-master-0.redis-headless.skupper-local.svc.cluster.local ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 62ms
rtt min/avg/max/mdev = 0.088/0.093/0.099/0.011 ms

sh-5.0# ping redis-headless-proxy-0.redis-headless.skupper-local.svc.cluster.local
ping: redis-headless-proxy-0.redis-headless.skupper-local.svc.cluster.local: Name or service not known

**Work around to  get this working**

I created a new service called **redis-headless-proxy** and updated the stateful set redis-headless-proxy to use the new service created. This will give the ability to get the pod identity for redis-headless-proxy pods.

$oc explain sts.spec.serviceName
KIND:     StatefulSet
VERSION:  apps/v1

FIELD:    serviceName <string>

DESCRIPTION:
     serviceName is the name of the service that governs this StatefulSet. This
     service must exist before the StatefulSet, and is responsible for the
     network identity of the set. Pods get DNS/hostnames that follow the
     pattern: pod-specific-string.serviceName.default.svc.cluster.local where
     "pod-specific-string" is managed by the StatefulSet controller.

### PR DESCRIPTION
**Pull request by grs**
Fixes #829

### CODE REVIEW

File: cmd/service-controller/controller.go [hunk 1]
```diff
@@ -655,20 +681,18 @@ func (c *Controller) processNextEvent() bool {
...
-					} else {
+					bindings, deleteSvc := c.getBindingsForService(svc)
+					if bindings != nil {
 						// check that service matches binding def, else update it
 						err = bindings.RealiseIngress()
 						if err != nil {
 							return err
 						}
+					} else if deleteSvc {
+						err = c.DeleteService(svc)
```
[the remark this answers is not in the document]
Comment (grs) [reply cmd/service-controller/controller.go:692]: I don't know. It didn't before. If it should I would suggest that is a separate fix.

[the remark this answers is not in the document]
Comment (nluaces) [reply cmd/service-controller/controller.go:692]: @fgiorgetti oh, I think I know what you mean; enabling/disabling tls support only occurs when the definition of services changes in the config map.

**Review by nluaces — APPROVED** [review]

**Review by fgiorgetti — APPROVED** [review]
The fix worked for me.
I was able to resolve the name for internal proxy pods using the new "-proxy" service.

**Review by ajssmith — APPROVED** [review]
lgtm, thanks!

**Pull request merged**